server/chat-plugins/teams.ts

```diff
@@ -0,0 +1,686 @@
+		` <i class="fa fa-refresh"></i> ${context.tr('Refresh')}</button>`
+	);
+}
+
+export const TeamsHandler = new class {
+	database = new PostgresDatabase();
+	readyPromise: Promise<void> | null = Config.usepostgres ? (async () => {
+		try {
+			await this.database.query('SELECT * FROM teams LIMIT 1');
+		} catch {
+			await this.database.query(FS(`databases/schemas/teams.sql`).readSync());
+		}
+	})() : null;
+	destroy() {
+		void this.database.destroy();
+	}
+
+	async search(search: TeamSearch, user: User, count = 10, includePrivate = false) {
+		const args = [];
+		const where = [];
+		if (search.format) {
+			where.push(`format = $${args.length + 1}`);
+			args.push(toID(search.format));
+		}
+		if (search.owner) {
+			where.push(`ownerid = $${args.length + 1}`);
+			args.push(toID(search.owner));
+		}
+		if (search.gen) {
+			where.push(`format LIKE 'gen${search.gen}%'`);
+		}
+		if (!includePrivate) where.push('private != true');
+
+		const result = await this.query<StoredTeam>(
+			`SELECT * FROM teams${where.length ? ` WHERE ${where.join(' AND ')}` : ''} LIMIT ${count}`,
+			args,
+		);
+		return result.filter(row => {
+			const team = Teams.unpack(row.team)!;
+			if (row.private && !(row.ownerid === user.id || user.can('rangeban'))) {
+				return false;
+			}
+			let match = true;
+			if (search.pokemon?.length) {
+				match = search.pokemon.some(
+					pokemon => team.some(set => toID(set.species) === toID(pokemon))
+				);
+			}
+			if (!match) return false;
+			if (search.moves?.length) {
+				match = search.moves.some(
+					move => team.some(set => set.moves.some(m => toID(m) === toID(move)))
+				);
+			}
+			if (!match) return false;
+			if (search.abilities?.length) {
+				match = search.abilities.some(
+					ability => team.some(set => toID(set.ability) === toID(ability))
+				);
+			}
+			return match;
+		});
+	}
+
+	async query<T = any>(statement: string, values: any[] = []) {
+		if (this.readyPromise) await this.readyPromise;
+		return this.database.query(statement, values) as Promise<T[]>;
+	}
+
+	async save(
+		connection: Connection,
+		formatName: string,
+		rawTeam: string,
+		teamName: string | null = null,
+		isPrivate = false,
+		isUpdate?: number
+	) {
+		this.validateAccess(connection, true);
+
+		if (Monitor.countPrepBattle(connection.ip, connection)) {
+			return null;
+		}
+		const user = connection.user;
+		const format = Dex.formats.get(toID(formatName));
+		if (!format.exists) {
+			connection.popup("Invalid format:\n\n" + formatName);
+			return null;
+		}
+		let existing = null;
+		if (isUpdate) {
+			existing = await this.get(isUpdate);
+			if (!existing) {
+				connection.popup("You're trying to edit a team that doesn't exist.");
+				return null;
+			}
+		}
+
+		const validatorResult = await TeamValidatorAsync.get(`gen${Dex.gen}customgame`).validateTeam(rawTeam, {
+			removeNicknames: true, // less headache to moderate this way
```

Comment:
Why? Aren't nicknames filtered anyway?

Reply:
I don't really think it's necessary to have nicknames in uploaded teams (how is that crucial to the set?) but i'm open to changing my mind on that one.

Reply:
It's not crucial information to a set but it provides flavor since some players have standard nicknames they give to specific Pokemon and whatnot. I just don't see the harm in allowing nicknames personally since they're already chat filtered.

Reply:
My thing is like, if we allow nicknames, then staff have to moderate this. And I don't think anyone wants to do that.